### ISSUE TITLE
Make "->"  works in all for-loop  to replace "do yield"

### ISSUE THREAD

**Issue by syeerzy** [fsharp/fslang-suggestions#764]
# Make "->"  works in all for-loop  to replace "do yield"

I propose we should had the same way to use "->" in all  for-loop segments

Now : 
for X in XXX do yield YYY      ======>    for X in XXX -> YYY      (OK)
for X=x to y do yield YYY      ======>    for X=x to y -> YYY      (Syntax Error)
for X=x downto y do yield YYY      ===>    for X=x downto y -> YYY      (Syntax Error)

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

**Comment by Happypig375** [comment]
Those two loops are considered "legacy" and should be discouraged, though.

**Comment by dsyme** [comment]
> Those two loops are considered "legacy" and should be discouraged, though.

Yes, though I suppose this means we should consider giving deprecation warnings too

**Comment by cartermp** [comment]
I don't really see that as being too good of a route. I'm not sure it's uniformly accepted that the equivalent Array/List/etc combinators are the way forward. Those can also be a bit more annoying to type out since you have to play the game of aligning the ending parenthesis.

**Comment by cartermp** [comment]
Generally I think this should be done for completeness' sake, even though we want people to use the collection functions instead.

**Comment by dsyme** [comment]
> I don't really see that as being too good of a route. I'm not sure it's uniformly accepted that the equivalent Array/List/etc combinators are the way forward. Those can also be a bit more annoying to type out since you have to play the game of aligning the ending parenthesis.

The proposition is that these two forms:

```
for X=x to y do ...
for X=x downto y do ...
```
should be deprecated in favour of
```
for X in a .. b do...
for X in a .. -1 .. b do...
```

**Comment by cartermp** [comment]
Where is this proposed?

**Comment by dsyme** [comment]
> Where is this proposed? 

Oh that's what I'm suggesting is the natural way to interpret this suggestion. I'm not  really pushing it - just saying it would be reasonable to completely remove the `for i=x to y do ...` and `for i = x downto y` from the language, since they are redundant (and the bug report is pointing out that these expression forms can't be used with `->` in sequence expressions.

These two looping constructs are only in F# because of OCaml (the `for i=x to y do ...` and `for i=x downto y do ...` are present in OCaml whereas `for i in x .. y do ...`  is not, it is an F# invention - partly because neither IEnumerable/seq nor the range construct `x .. y` are present in OCaml)

> Also, this would represent a regression, since you can't have the index while you're doing that.

The loops `for i in x .. y do` and `for i in x .. -1 .. y` are still index loops over ranges of integers, the forms are equivalent.  Also both are also optimized to C-style for-loops.

**Comment by cartermp** [comment]
Hmmm. Okay, perhaps. I just feel like I'd rather focus on deprecating things that are problematic but possible (e.g, `use mutable x = ...`), not necessarily things that just feel like they have legacy.

**Comment by dsyme** [comment]
Given that `yield` is now no longer needed in most situations, I think this is no longer compelling, as you can just use `do` everywhere.
